src/js/experiments/snippetFeedback.js

```diff
@@ -0,0 +1,250 @@
+
+        document.addEventListener('click', outsideClickListener);
+      }, 0);
+    }
+  }
+
+  /**
+   * Render the char count and set it's result on the respective element.
+   * */
+  renderCharCount(charCount) {
+    let charCountLimited =
+      charCount > this.maxCharCount ? this.maxCharCount : charCount;
+    const el = `
+        <div class="charCountRow">
+          <span>${charCountLimited} / ${this.maxCharCount}</span>
+        </div>`;
+    this.currCharCountElement.innerHTML = el;
+  }
+
+  // -- Trackers --
+  // These are responsible for sending data based on user input
+  // to amplitude.
+
+  /**
+   * trackYesOrNoButton sends a trackAction to aplitude.
+   * actions are: "docs-snippet-helpful-no"  and "docs-snippet-helpful-yes"
+   * */
+  trackYesOrNoButton(yesOrNoString) {
+    window.AnalyticsClient.trackAction(
+      `docs-snippet-helpful-${yesOrNoString}`,
+      {
+        originatingSnippet: this.snippetElement.textContent,
+        location: window.location.pathname,
+        timeOfButtonClick: Date.now(),
+      },
+    );
+  }
+
+  /**
+   * `trackFormSubmission` invokes trackAction with a payload of the original
+   * snippet, the feedback of the user, and the timeOfFormSubmission.
+   * */
+  trackFormSubmission(formContent) {
+    window.AnalyticsClient.trackAction(`docs-snippet-helpful-form-submission`, {
+      originatingSnippet: this.snippetElement.textContent,
+      feedback: formContent,
+      location: window.location.pathname,
+      wasThisHelpful: this.wasThisHelpful,
+      // for diffing this against the time that the trackYesOrNoButton was clicked
+      // as well as possible the time the copy code button was clicked (TODO add that)
+      timeOfFormSubmission: Date.now(),
+    });
+  }
+
+  // -- Helpers --
+
+  /**
+   * Construct a dom element, setting classes, text content, class, onclick etc.
+   * */
+  makeElement({ kind, text, className, onClick }) {
+    let el = document.createElement(kind);
+    if (text) el.textContent = text;
+    if (className) el.classList.add(className);
+    if (onClick) el.addEventListener('click', onClick);
+
+    return el;
+  }
+}
+
+/**
+ * init queries the dom for all snipppets
+ * TODO(blocked): when that snippet's COPY CODE button is clicked... (waiting for graduation of copy code buttons)
+ *
+ * */
+function init() {
+  // add event listeners to all code snippets on page
+  window.OptimizelyClient.getVariationName({
+    experimentKey: 'dd_docs_snippet_feedback_experiment_test',
+    groupExperimentName: 'q4_fy22_docs_disco_experiment_group_test',
+    experimentContainer: '.hljs',
+  }).then((variation) => {
+    if (variation === 'treatment') {
+      const snippets = document.querySelectorAll('code.hljs');
+      [...snippets].forEach((snippetBlock) => {
+        snippetBlock.addEventListener(
+          'click',
+          (pointerEvent) => {
+            let snippetElement = pointerEvent.target;
+            new SnippetFeedback(snippetElement);
```

Comment:
we should probably be passing the index of the snippetElement so we can track back the index of the code snippet.

Reply:
do you think we should change the logic in highlight js to be the same? these should ideally track in the same way.

no payload restriction but is there a cost associated with storing the entire code snippet repeatedly? I'm not familiarly with the pricing of our Amplitutude & Segment. @rsylvian can you chime in?

I'm not sure how I feel about CTRL+F to find a location/page of a snippet in our code base. Content changes are a valid concern but I think we can see when changes were made in git history to understand our charts.

Reply:
We should not store the entire code snippet. While all the above concerns are valid, having an entire code snippet in the metric is not going to produce data/charts that are usable according to DA. Please store location instead 🙏